contracts-bedrock: move immutables to storage for L1ERC721Bridge

0xfuturistic opened the pull request:
context: https://github.com/ethereum-optimism/client-pod/issues/265
security review (wip): https://www.notion.so/oplabs/DRAFT-Security-Review-Move-Immutables-to-Storage-in-L1-Contracts-56c07789fa454d8fb2df80f0b84a483b?pvs=4
legacy PR: https://github.com/ethereum-optimism/optimism/pull/6477

0xfuturistic committed:
contracts-bedrock: move immutables to storage in ERC721Bridge

0xfuturistic committed:
contracts-bedrock: update L1ERC721Bridge to support changes to ERC721Bridge

0xfuturistic committed:
contracts-bedrock: update L2ERC721Bridge to support changes to ERC721Bridge

0xfuturistic committed:
contracts-bedrock: update Deploy script to support changes to L1ERC721Bridge

L1ERC721Bridge had its immutables moved to storage

0xfuturistic committed:
contracts-bedrock: remove requirements in __ERC721Bridge_init

This is needed because a call to __ERC721Bridge_init is triggered by the constructors of L1ERC721Bridge and L2ERC721Bridge. However, since the variables required to be non-null by __ERC721Bridge_init are set to null values in those constructors, this will call revert and the set-ups of L1ERC721Bridge and L2ERC721Bridge will fail. This commit fixes that issue.

0xfuturistic committed:
contracts-bedrock: fix Deploy script for L1ERC721Bridge

Since values are set at initialization and not at deployment, ChainAssertions.checkL1ERC721Bridge reverts if called at this point with the dependent contracts. Instead, this call should be done immediately after initialization (as is already done in the Deploy script).

0xfuturistic committed:
op-chain-ops: update upgrade script for L1ERC721Bridge

L1ERC721Bridge had its immutable variables moved to storage

0xfuturistic committed:
contracts-bedrock: fix L1ERC721Bridge spec in Authorization.t.sol

Addresses the fact that L1ERC721Bridge spec for initialize should include the parameters for the variables moved from immutables into storage

0xfuturistic committed:
contracts-bedrock: update semver for L1ERC721Bridge, L2ERC721Bridge

0xfuturistic committed:
op-chain-ops: redefine L2ERC721Bridge immutable config

immutable variables were removed

0xfuturistic committed:
op-chain-ops: introduce storage values for L2ERC721Bridge

corresponding to the immutable variables moved into storage in the contract

0xfuturistic committed:
op-chain-ops: relabel variables in L1ERC721Bridge upgrade script

0xfuturistic committed:
contracts-bedrock: fix annotation in ERC721Bridge

0xfuturistic committed:
contracts-bedrock: reintroduce ChainAssertions for L1ERC721Bridge at deploy time

0xfuturistic committed:
contracts-bedrock: reintroduce parameter to L2ERC721Bridge constructor

this makes the changes backwards compatible with existing deploy scripts

0xfuturistic committed:
op-chain-ops: fix bug, missing parameter for L2ERC721Bridge deploy

0xfuturistic committed:
contracts-bedrock: make snapshots for L1ERC721Bridge, L2ERC721Bridge

0xfuturistic committed:
contracts-bedrock: update semver-lock

0xfuturistic committed:
contracts-bedrock: move repeated code in ChainAssertions's checkL1ERC721Bridge

repeated code is moved outside the if-else block

tynes commented on .abigenrc:
```diff
@@ -1 +1 @@
-v1.10.25
+v1.10.16
```
Why is this diff here?

0xfuturistic replied:
downgraded to a version of abigen Matt recommended me (bindings weren't matching those of CI), which was eventually saved to the file after running `pnpm upgrade:abigen` in repo root. can revert back though

tynes replied:
Would prefer to keep the same version of abigen if possible. If it results in 100% issues with matching bindings then we can figure out why. This sometimes happened to me with weth9

clabby replied:
I imagine this was a mismatch with the foundry version, recently there was a change in how the ABI items are sorted. Try going back to the old version of abigen, and `p update:foundry` in the monorepo root.

0xfuturistic committed:
contracts-bedrock: replace isInitialized by isProxy in ChainAssertions for L1ERC721Bridge

0xfuturistic committed:
op-chain-ops: fix bug in L1ERC721Bridge upgrade script

0xfuturistic committed:
contracts-bedrock: update gap in ERC721Bridge

done to maintain existing storage layout

0xfuturistic committed:
contracts-bedrock: update snapshots for L1ERC721Bridge, L2ERC721Bridge

0xfuturistic committed:
contracts-bedrock: update semver-lock

0xfuturistic committed:
op-bindings: make for L1ERC721Bridge, L2ERC721Bridge

0xfuturistic committed:
contracts-bedrock: update semver-lock

0xfuturistic committed:
contracts-bedrock: change !isProxy isProxy order in ChainAssertions for L1ERC721Bridge

0xfuturistic committed:
contracts-bedrock: add tests for L1ERC721Bridge

0xfuturistic committed:
contracts-bedrock: add tests for L2ERC721Bridge

0xfuturistic committed:
contracts-bedrock: fix test for L1ERC721Bridge

0xfuturistic committed:
op-chain-ops: reorder setting of storage variables in config

makes the order consistent with previous order for immutables

0xfuturistic committed:
contracts-bedrock: use address from predeploys in Deploy script for L1ERC721Bridge

0xfuturistic committed:
op-chain-ops: check initialize of L2ERC721Bridge

0xfuturistic committed:
contracts-bedrock: add require and tests for ERC721Bridge

0xfuturistic committed:
contracts-bedrock: update semver-lock

0xfuturistic committed:
op-bindings: make for L1ERC721Bridge, L2ERC721Bridge

tynes commented on op-chain-ops/upgrades/l1.go:
```diff
@@ -168,7 +169,26 @@ func L1ERC721Bridge(batch *safe.Batch, implementations superchain.Implementation
 
-	calldata, err := l1ERC721BridgeABI.Pack("initialize", superchainConfigProxy)
+	var messenger, otherBridge common.Address
+	if config != nil {
```
Confused about this check as `config` isn't used in the branch, apologies if it was previously implemented this way

tynes replied:
We should pull the remote values and compare them against the config if the config exists and error if they are different

tynes commented on packages/contracts-bedrock/src/L1/L1ERC721Bridge.sol:
```diff
@@ -23,19 +23,32 @@ contract L1ERC721Bridge is ERC721Bridge, ISemver {
+    constructor() ERC721Bridge() {
+        initialize({
+            _messenger: CrossDomainMessenger(address(0)),
+            _otherBridge: Predeploys.L2_ERC721_BRIDGE,
```
We should set `_otherBridge` to `address(0)` to make the singleton completely broken

tynes commented on packages/contracts-bedrock/src/L1/L1ERC721Bridge.sol:
```diff
@@ -23,19 +23,32 @@ contract L1ERC721Bridge is ERC721Bridge, ISemver {
+        initializer
+    {
         superchainConfig = _superchainConfig;
+        __ERC721Bridge_init(_messenger, _otherBridge);
```
nit: can you update the call to `__ERC721Bridge_init` to using kwargs style syntax, it is easier to review. Will guarantee the args are in the correct order

tynes commented on packages/contracts-bedrock/src/L1/L1ERC721Bridge.sol:
```diff
@@ -110,7 +123,7 @@ contract L1ERC721Bridge is ERC721Bridge, ISemver {
 
         // Send calldata into L2
-        MESSENGER.sendMessage(OTHER_BRIDGE, message, _minGasLimit);
+        messenger.sendMessage(otherBridge, message, _minGasLimit);
```
Can you also update this to using kwargs style syntax for both `sendMessage` and `transferFrom`? Refactoring as we touch code to improve code quality, our style guide has since updated to prefer this style especially when there is more than 2 args

tynes commented on packages/contracts-bedrock/src/L2/L2ERC721Bridge.sol:
```diff
@@ -20,18 +20,20 @@ import { Predeploys } from "src/libraries/Predeploys.sol";
-    constructor(address _otherBridge) ERC721Bridge(Predeploys.L2_CROSS_DOMAIN_MESSENGER, _otherBridge) {
-        initialize();
+    constructor(address _otherBridge) ERC721Bridge() {
+        initialize({ _messenger: CrossDomainMessenger(Predeploys.L2_CROSS_DOMAIN_MESSENGER), _otherBridge: _otherBridge });
```
We want to hardcode `address(0)` in the constructor for these so the singleton is dead

0xfuturistic replied:
should the parameter for the constructor remain? to minimize changes to L2 contracts, I didn't change it. but happy to do so in light of making the singleton as dead as possible (i.e., setting `address(0)` in the constructor for otherBridge rather than taking a parameter)

tynes commented on packages/contracts-bedrock/src/L2/L2ERC721Bridge.sol:
```diff
@@ -20,18 +20,20 @@ import { Predeploys } from "src/libraries/Predeploys.sol";
+    /// @notice Initializes the contract.
+    /// @param _messenger   Contract of the CrossDomainMessenger on this network.
+    /// @param _otherBridge Address of the ERC721 bridge on the other network.
+    function initialize(CrossDomainMessenger _messenger, address _otherBridge) public initializer {
```
We can hardcode the `CrossDomainMessenger` to be the predeploy address here and only pass `_otherBridge` as the argument

tynes commented on packages/contracts-bedrock/src/L2/L2ERC721Bridge.sol:
```diff
@@ -20,18 +20,20 @@ import { Predeploys } from "src/libraries/Predeploys.sol";
+    /// @param _messenger   Contract of the CrossDomainMessenger on this network.
+    /// @param _otherBridge Address of the ERC721 bridge on the other network.
+    function initialize(CrossDomainMessenger _messenger, address _otherBridge) public initializer {
+        __ERC721Bridge_init(_messenger, _otherBridge);
```
nit: kwargs style syntax

tynes commented on packages/contracts-bedrock/src/L2/L2ERC721Bridge.sol:
```diff
@@ -113,7 +115,7 @@ contract L2ERC721Bridge is ERC721Bridge, ISemver {
         // Send message to L1 bridge
         // slither-disable-next-line reentrancy-events
-        MESSENGER.sendMessage(OTHER_BRIDGE, message, _minGasLimit);
+        messenger.sendMessage(otherBridge, message, _minGasLimit);
```
nit: kwargs style syntax

tynes commented on packages/contracts-bedrock/src/universal/ERC721Bridge.sol:
```diff
@@ -57,32 +55,36 @@ abstract contract ERC721Bridge is Initializable {
-        MESSENGER = CrossDomainMessenger(_messenger);
-        OTHER_BRIDGE = _otherBridge;
+        messenger = _messenger;
+        otherBridge = _otherBridge;
```
We should be able to hardcode `_otherBridge` to be the predeploy constant here

tynes replied:
Wait nevermind this is the base impl not the L1 impl

0xfuturistic committed:
contracts-bedrock: set otherBridge to null address in constructor of L1ERC721Bridge

0xfuturistic committed:
contracts-bedrock: hardcode messenger in L2ERC721Bridge to Predeploys

0xfuturistic committed:
contracts-bedrock: drop require for bridge to be null in ERC721Bridge

0xfuturistic committed:
op-bindings: make for L1ERC721Bridge, L2ERC721Bridge

0xfuturistic committed:
contracts-bedrock: update semver-lock

0xfuturistic committed:
op-chain-ops: fix L2ERC721Bridge initialize parameter count

tynes commented on op-chain-ops/cmd/check-l2/main.go:
```diff
@@ -563,6 +563,34 @@ func checkL2ERC721Bridge(addr common.Address, client *ethclient.Client) error {
+		return err
+	}
+
+	initializing, err := getInitializing("L2ERC721Bridge", addr, client)
+	if err != nil {
+		return err
+	}
+	log.Info("L2ERC721Bridge", "_initializing", initializing)
```
This is a good check, we want initializing to be false in all cases

tynes commented on packages/contracts-bedrock/src/L1/L1ERC721Bridge.sol:
```diff
@@ -23,19 +23,32 @@ contract L1ERC721Bridge is ERC721Bridge, ISemver {
+    /// @param _superchainConfig Contract of the SuperchainConfig contract on this network.
+    function initialize(
+        CrossDomainMessenger _messenger,
+        address _otherBridge,
```
We will always know the address of `_otherBridge` on L1 - its the L2 predeploy. We can remove the argument here and hardcode the arg to `__ERC721Bridge_init`

tynes commented on packages/contracts-bedrock/src/L2/L2ERC721Bridge.sol:
```diff
@@ -20,18 +20,22 @@ import { Predeploys } from "src/libraries/Predeploys.sol";
-    /// @param _otherBridge Address of the ERC721 bridge on the other network.
-    constructor(address _otherBridge) ERC721Bridge(Predeploys.L2_CROSS_DOMAIN_MESSENGER, _otherBridge) {
-        initialize();
+    constructor(address _otherBridge) ERC721Bridge() {
```
We don't need an argument here - it should hardcode `address(0)` in the singleton

tynes commented on packages/contracts-bedrock/src/universal/ERC721Bridge.sol:
```diff
@@ -9,18 +9,16 @@ import { Initializable } from "@openzeppelin/contracts/proxy/utils/Initializable
-    address public immutable OTHER_BRIDGE;
+    /// @notice Address of the bridge on the other network.
+    /// @custom:network-specific
+    address public otherBridge;
```
Can you make this type be a `StandardBridge` for consistency?

0xfuturistic committed:
contracts-bedrock: change type of otherBridge in ERC721Bridge

0xfuturistic committed:
contracts-bedrock: drop constructor arg, use Predeploy in L1ERC721Bridge

0xfuturistic committed:
contracts-bedrock: update Deployer's loadInitializedSlot

if contract is a proxy, it now gets the implementation address using either EIP1967Helper or AddressManager (for ResolvedDelegateProxy)

0xfuturistic committed:
contracts-bedrock: improve logic in loadInitializedSlot

0xfuturistic committed:
contracts-bedrock: use solady LibString in loadInitializedSlot

0xfuturistic committed:
contracts-bedrock: define constants in EIP1967Helper and drop Constants import

this allows us to not have to update the version of Arithmetic, ResourceMetering, and Burn, which would otherwise clash with Deployer.sol

0xfuturistic committed:
contracts-bedrock: use interface for AddressManager instead of updating version directly

0xfuturistic committed:
op-bindings: make

0xfuturistic committed:
contracts-bedrock: drop redundant tests

0xfuturistic committed:
contracts-bedrock: fix rebase

0xfuturistic committed:
op-chain-ops: check fetched addresses from OptimismPortal against DeployConfig

0xfuturistic committed:
contracts-bedrock: update semver-lock

0xfuturistic committed:
contracts-bedrock: make snapshots

tynes commented on op-chain-ops/upgrades/l1.go:
```diff
@@ -168,7 +169,30 @@ func L1ERC721Bridge(batch *safe.Batch, implementations superchain.Implementation
+			return fmt.Errorf("upgrading L1ERC721Bridge: Messenger address doesn't match config")
+		}
+
+		if otherBridge != predeploys.L2ERC721BridgeAddr {
```
This doesn't need to be in the `config != nil` branch

tynes commented on op-chain-ops/upgrades/l1.go:
```diff
@@ -168,7 +169,30 @@ func L1ERC721Bridge(batch *safe.Batch, implementations superchain.Implementation
+		}
+	}
+
+	calldata, err := l1ERC721BridgeABI.Pack("initialize", messenger, otherBridge, superchainConfigProxy)
```
This call here doesn't need `otherBridge`

tynes commented:
Good for merge after https://github.com/ethereum-optimism/optimism/pull/8633#discussion_r1448083778 is addressed

0xfuturistic committed:
op-chain-ops: drop unnecessary parameter for l1ERC721BridgeABI in L1 upgrades

0xfuturistic committed:
op-chain-ops: move check out of restrictive clause in L1 upgrade for L1ERC721Bridge

tynes commented:
Looks like merge conflict

0xfuturistic committed:
contracts-bedrock: fix after rebase Initializable

0xfuturistic committed:
op-chain-ops: drop redundant import in L1 upgrade
